airbyte-integrations/connectors/source-shopify/source_shopify/shopify_graphql/bulk/query.py

```diff
@@ -0,0 +1,184 @@
+                        ... on BulkOperation {
+                            id
+                            status
+                            errorCode
+                            objectCount
+                            fileSize
+                            url
+                            partialDataUrl
+                        }
+                    }
+                }"""
+        ).substitute(job_id=bulk_job_id)
+
+    @staticmethod
+    def prepare(query: str) -> str:
+        bulk_template = Template(
+            '''mutation {
+                bulkOperationRunQuery(
+                    query: """
+                    $query
+                    """
+                ) {
+                    bulkOperation {
+                        id
+                        status
+                    }
+                    userErrors {
+                        field
+                        message
+                    }
+                }
+            }'''
+        )
+        return bulk_template.substitute(query=query)
+
+
+class GraphQlQueryBuilder:
+
+    operation = "query"
+    edge_key = "edges"
+    node_key = "node"
+
+    def get_edge(self, name: str, fields: Union[List[str], List[Field], str]) -> Field:
+        """
+        Defines the edge of the graph and it's fields to select.
+        """
+        return Field(name=name, fields=[Field(name=self.edge_key, fields=[Field(name=self.node_key, fields=fields)])])
+
+    def get_query(self, name: str, edges: Union[List[Field], Field] = None, filter_query: str = None, sort_key: str = None) -> Query:
+        """
+        Defines the root of the graph with edges.
+        """
+        args: List[Argument] = []
+        # constructing arguments
+        if filter_query:
+            args.append(Argument(name=self.operation, value=f'"{filter_query}"'))
+        if sort_key:
+            args.append(Argument(name="sortKey", value=sort_key))
+        # constructing edges
+        fields = [
+            Field(name=self.edge_key, fields=[Field(name=self.node_key, fields=["id", edges] if edges else ["id"])]),
+        ]
+        # return constucted query
+        return Query(name=name, arguments=args, fields=fields)
+
+
+class ShopifyBulkQuery(GraphQlQueryBuilder):
+    @property
+    @abstractmethod
+    def record_identifier(self) -> str:
+        """
+        Defines the record identifier to fetch only records related to the choosen stream.
+        Example:
+            { "admin_graphql_api_id": "gid://shopify/Metafield/22533588451517" }
+            In this example the record could be identified by it's reference = ".../Metafield/..."
+        The property should be defined like:
+            record_identifier = "Metafield"
+        """
+
+    @property
+    @abstractmethod
+    def graph_edge(self) -> str:
+        """
+        Defines the final graph node name to fetch from.
+        """
+
+    @property
+    @abstractmethod
+    def graph_fields(self) -> Union[List[str], List[Any]]:
```

Comment:
why can this be a `List[Any]`? can we restrict this a little more? looking at the current usage, I think `List[str]` should be enough